andi/andi.py

```diff
@@ -265,24 +270,54 @@ def build(plan):  # Build all the instances from a plan
         In other words, the kwargs dict returned by the method
         ``Plan.final_kwargs`` could not contain all required
         arguments to build/invoke the input class/function.
+    :param overrides: A funtion that maps a class/function to a different
+        class/function. The function must return None if no special mapping should
+        be applied for a particular class/function. The suggested remapping serves
+        to override a particular dependency in the dependency tree, and
+        update the plan acordingly. For example, you might want to replace class
+        ``PenneWithTomate`` by ``SpaghettiBolognese`` whenever it is find in the
+        dependendy tree, but updating the plan so that ``SpaghettiBolognese``
+        gets its dependencies ``Meat`` and ``Spaghetti`` resolved properly.
+    :param recursive_overrides: If True, ``overrides`` are applied recursively
```

Comment:
I wonder if there are use cases when some classes need to be overridden recursively, while others not, in the same plan. What do you think about supporting it? It'd likely mean there is no `recursive_overrides` option, and  e.g. that `overrides` function should return a flag to indicate if an override for a class is final.

Reply:
Let's see an example:

```py 
class DBConn(Injectable):
  ...

attr.s(auto_attribs=True)
class GenerateWebPage(Injectable):
  """Generates HTML given data got from the database"""
  conn: DBConn

  def to_html():
    data = self.conn.query(...)
    # Do something with data and generate html
    return ...

attr.s(auto_attribs=True)
class RefineGeneratedWebPage(Injectable):
  conn: DBConn
  generator :GenerateWebPage

  def to_html():
     html = self.generator.to_html()
     new_data = conn.query(...)
     # refine the html given the new data
     return html
```

We want `DBConn` to be overridden in any case, as we might change the database connection to `MysqlDBConn` or `PosgresDBConn` for example. But we don't want recursivity for `RefineGeneratedWebPage` wrt. `GenerateWebPage`. 

Note that the proposed idea of a flag that marks an override as final won't cover this case. If the `GenerateWebPage` is marked as final, then the database connection won't be overridden, but this is not what is desired. 

So my conclusion is that it might be interesting to be able to customize more the overrides, but the configuration wouldn't be as simple as a final flag. Two ideas:
- Provide the pairs of types for which no recursive overrides is allowed (I don't like much)
- Instead of stopping overrides for any types when an override happened, what if we just stop overrides for the overridden type? The rest of the overrides would keep working. It sounds good, but I think it will have secondary effects (e.g. if a non-affected dependency finally depends on overridden type). 

Also, I'm not sure we should think about all this as this stage. Probably we should wait for the real need to think on that, as I'm not sure this is really needed.

Reply:
Ok, let's wait for a real need :) I was looking at it mostly from the API point of view, that `recursive_overrides` parameter may not be needed if this flag is per-class we're overriding.